ui/helpers/formatters.ts

```diff
@@ -0,0 +1,143 @@
+import { useSelector } from 'react-redux';
+import { getResolvedLocale } from '../ducks/locale/locale';
+
+// TODO: Move this to @metamask/core
+
+const twoDecimals = {
+  minimumFractionDigits: 2,
+  maximumFractionDigits: 2,
+};
+
+const threeDecimals = {
+  minimumFractionDigits: 3,
+  maximumFractionDigits: 3,
+};
+
+const compactTwoDecimals: Intl.NumberFormatOptions = {
+  notation: 'compact',
+  maximumFractionDigits: 2,
+};
+
+const numberFormatCache = new Map<string, Intl.NumberFormat>();
+
+function getCachedNumberFormat(
+  locale: string,
+  options: Intl.NumberFormatOptions = {},
+) {
+  const key = `${locale}_${JSON.stringify(options)}`;
+
+  if (!numberFormatCache.has(key)) {
+    try {
+      numberFormatCache.set(key, new Intl.NumberFormat(locale, options));
+    } catch (error) {
+      if (error instanceof RangeError) {
+        // Fallback for invalid options (e.g. currency code)
+        numberFormatCache.set(key, new Intl.NumberFormat(locale, twoDecimals));
+      } else {
+        throw error;
+      }
+    }
+  }
+
+  // Do we need to update TypeScript ESLint to prevent this false positive?
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+  return numberFormatCache.get(key)!;
+}
+
+function formatCurrency(
+  config: { locale: string },
+  value: number | bigint | `${number}`,
+  currency: Intl.NumberFormatOptions['currency'],
+  options: Intl.NumberFormatOptions = {},
+) {
+  if (!Number.isFinite(Number(value))) {
+    return '';
+  }
+
+  const numberFormat = getCachedNumberFormat(config.locale, {
+    style: 'currency',
+    currency,
+    ...options,
+  });
+  return numberFormat.format(value);
+}
+
+function formatCurrencyCompact(
+  config: { locale: string },
+  value: number | bigint | `${number}`,
+  currency: Intl.NumberFormatOptions['currency'],
+) {
+  return formatCurrency(config, value, currency, compactTwoDecimals);
+}
+
+function formatCurrencyWithThreshold(
+  config: { locale: string },
+  value: number | bigint | `${number}`,
+  currency: Intl.NumberFormatOptions['currency'],
+) {
+  if (!Number.isFinite(Number(value))) {
+    return '';
+  }
+
+  const number = Number(value);
+  let options: Intl.NumberFormatOptions = {};
+
+  if (number === 0) {
+    return formatCurrency(config, 0, currency);
+  } else if (number < 0.001) {
+    const formatted = formatCurrency(config, 0.001, currency, threeDecimals);
+    return `<${formatted}`;
+  } else if (number < 0.01) {
+    const formatted = formatCurrency(config, 0.01, currency, twoDecimals);
+    return `<${formatted}`;
+  } else if (number < 1) {
+    options = threeDecimals;
+  } else if (number < 1e6) {
+    options = {};
+  } else if (number < 1e15) {
+    options = compactTwoDecimals;
+  }
+
+  return formatCurrency(config, number, currency, options);
+}
```

Comment:
Which currencies has 3 decimal places? Are we planning to use this function to format fiats and cryptos?